registry/storage/driver/azure/azure.go

```diff
@@ -399,38 +372,48 @@ func directDescendants(blobs []string, prefix string) []string {
 	return keys
 }
 
-func (d *driver) listBlobs(container, virtPath string) ([]string, error) {
+func (d *driver) listBlobs(ctx context.Context, virtPath string) ([]string, error) {
 	if virtPath != "" && !strings.HasSuffix(virtPath, "/") { // containerify the path
 		virtPath += "/"
 	}
 
+	// we will replace the root directory prefix before returning blob names
+	blobPrefix := d.blobName("")
+
+	// This is to cover for the cases when the rootDirectory of the driver is either "" or "/".
+	// In those cases, there is no root prefix to replace and we must actually add a "/" to all
+	// results in order to keep them as valid paths as recognized by storagedriver.PathRegexp
+	prefix := ""
+	if blobPrefix == "" {
+		prefix = "/"
+	}
+
 	out := []string{}
-	marker := ""
-	containerRef := d.client.GetContainerReference(d.container)
-	for {
-		resp, err := containerRef.ListBlobs(azure.ListBlobsParameters{
-			Marker: marker,
-			Prefix: virtPath,
-		})
-		if err != nil {
-			return out, err
-		}
 
-		for _, b := range resp.Blobs {
-			out = append(out, b.Name)
+	listPrefix := d.blobName(virtPath)
+	pager := d.client.NewListBlobsFlatPager(&container.ListBlobsFlatOptions{
+		Prefix: &listPrefix,
+	})
+	for pager.More() {
+		resp, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, err
 		}
-
-		if len(resp.Blobs) == 0 || resp.NextMarker == "" {
-			break
+		for _, blob := range resp.Segment.BlobItems {
+			name := *blob.Name
```

Comment:
This dereference needs a guard. It would make no sense for the API to return a response lacking a name, but remote APIs can have bugs.